Homework #1. Task #2. Волынец Арсений Викторович

vol0n opened the pull request.

vol0n committed:
Initial commit

vol0n committed:
Initial commit

nbirillo commented on src/main/kotlin/homework_1/task_2.kt:
```diff
@@ -0,0 +1,13 @@
+fun countOccurrences(where: String, what: String): Int =
```
Опять нет пакета. Плюс нужно ответвляться от мастера, а не от прошлой ветки. Так что поменяй исходную ветку, пожалуйста, чтобы тут (в этом пр) была только вторая задача.

vol0n replied:
Да, я сильно затупил с ветками. То есть я правильно понял, что сейчас HEAD направляю туда же, где и мастер, а потом создаю новую ветку, добавляю туда файл со исправленной второй задачей и делаю новый пр? Или можно как-то эту ветку ответвить и сделать потом просто сделать коммит и он после пуша будет в этом пр? Просто я перевелся и еще с гитом не особо подружился

nbirillo replied:
Ну первый вариант работать будет, но так делать нежелательно) Попробуй использовать `git rebase` для начала

nbirillo commented on src/main/kotlin/homework_1/task_2.kt:
```diff
@@ -0,0 +1,13 @@
+fun countOccurrences(where: String, what: String): Int =
+    where.substring(0, where.length - what.length + 1)
```
Если ты посмотришь на реализацию, то `substring` копирует строку, это плохо, так что давай поищем другой способ решить эту задачу)

vol0n replied:
Вроде я все просмотрел по доступу к кускам коллекции(substring, slice, chunked, windowed), там везде создаётся новый объект, как-то без циклов решить не получается. Мне пришло в голову только написать функцию, которая принимает индекс, строку и подстроку, и она сравнивает есть совпадает ли кусок с текущего индекса в строке с подстрокой, а потом с помощью foldIndexed посчитать сумму. 

fun compareChunk(idx: Int, source: String, substr: String): Boolean {
    var i = idx
    var j = 0
    while (i < substr.length){
        if (source[i] != substr[j])
            return false
        i++; j++
    }
    return true;
}

fun countOccurrences(where: String, what: String): Int =
    where.foldIndexed(0) {idx, count, _ ->
        count + if (idx < where.length - what.length + 1 && compareChunk(idx, where, what)) 1 else 0
    }

Неужели есть способ это как-то красиво без циклов сделать?)

nbirillo replied:
можешь посмотреть в сторону [windowed](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/windowed.html) например :)

nbirillo replied:
В котлине почти все можно сделать, не используя напрямую циклы

vol0n replied:
В windowed вроде List<String> возвращается, а с циклом можно без создания объектов обойтись)

nbirillo replied:
ну да, ну ты же попросил вариант как сделать красиво без циклов)

nbirillo replied:
но ты довольно неплохо тоже предложил на самом деле если что)

nbirillo commented on src/main/kotlin/homework_1/task_2.kt:
```diff
@@ -0,0 +1,13 @@
+    .mapIndexed { idx, _ -> if (where.substring(idx, idx + what.length) == what) 1 else 0 }
+    .reduce { sum, x -> sum + x }
+
+fun main() {
```
Если ввести две пустые строки, то будет ошибка

nbirillo replied:
Если первая строка пустая, а вторая нет - ошибка

nbirillo replied:
Если вторая строка пустая, а первая нет, то ошибка

vol0n replied:
А если подстрока пустая, то надо считать как в теории множеств, что пустое множество содержится в любом множестве?

nbirillo replied:
На твое усмотрение, главное обработать все случаи

vol0n committed:
Apply requested changes

-Теперь функция countOccurences работает без создания копий строки
-Учтены случаи когда одна (обе) строки могут быть пустыми
-Добавлено имя пакета
-Изменена настройка опции PackageNaming, чтобы допускались цифры в
названии пакета

vol0n committed:
Fix conflicts

vol0n committed:
Fix merge conflicts

vol0n committed:
Fix detekt FinalNewLine error

nbirillo commented on src/main/kotlin/homework1/task_2.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+}
+
+fun main() {
```
Решение довольно красивые было бы, если бы ты его починил :)
![image](https://user-images.githubusercontent.com/8993915/110080529-a1a0b980-7d9b-11eb-82a4-657f6c232bdb.png)

nbirillo commented on src/main/kotlin/homework1/task_2.kt:
```diff
@@ -0,0 +1,32 @@
+}
+
+@Suppress("ReturnCount")
+fun countOccurrences(where: String, what: String): Int {
```
Вообще в Котлине прикольно писать расширения, то есть в твоем случае:
```
fun String.countOccurrences(substr: String): Int { ... }
```

Это при вызове очень аккуратно и красиво читается и нельзя перепутать, что строка, а что подстрока. Это конечно дело вкуса, но часто выглядит симпатичнее обычных функций

vol0n committed:
Fix CountOccurences

nbirillo reviewed (approved):
зачтено
